sympy/core/expr.py

```diff
@@ -329,98 +329,55 @@ def __complex__(self):
         re, im = result.as_real_imag()
         return complex(float(re), float(im))
 
-    def __ge__(self, other):
-        from sympy import GreaterThan
+    def _cmp(self, other, op, cls):
+        assert op in ("<", ">", "<=", ">=")
         try:
             other = _sympify(other)
         except SympifyError:
-            raise TypeError("Invalid comparison %s >= %s" % (self, other))
+            raise TypeError("Invalid comparison %s %s %s" % (self, op, other))
         for me in (self, other):
             if me.is_extended_real is False:
                 raise TypeError("Invalid comparison of non-real %s" % me)
             if me is S.NaN:
                 raise TypeError("Invalid NaN comparison")
+
         n2 = _n2(self, other)
         if n2 is not None:
-            return _sympify(n2 >= 0)
+            return eval("_sympify(n2 %s 0)" % op)
+
         if self.is_extended_real and other.is_extended_real:
-            if (self.is_infinite and self.is_extended_positive) \
-                    or (other.is_infinite and other.is_extended_negative):
-                return S.true
-            nneg = (self - other).is_extended_nonnegative
-            if nneg is not None:
-                return sympify(nneg)
-        return GreaterThan(self, other, evaluate=False)
+            # this should work for infinite values as well,
+            # so we don't need a separate check
+            diff = self - other
+            if diff is not S.NaN:
+                test = eval("diff.%s" %
+                            {"<": "is_extended_negative",
+                             ">": "is_extended_positive",
+                             "<=": "is_extended_nonpositive",
```

Comment:
We shouldn't use eval here either. There might be a good way of doing this with lambda functions or with operator.attrgetter and operator.le etc.

Reply:
Just do

```py
if op == '<':
    test = diff.is_extended_negative
elif op == '>':
    test = diff.is_extended_positive
...
```

You lose the value of removing code duplication if you make the code harder to read and more bug prone.

Reply:
Okay, I changed it